File: .skills/openclaw-skills/skills/0xnyk/xint/lib/report.ts
/**
 * lib/report.ts — Automated intelligence reports.
 *
 * Combines search, sentiment analysis, and Grok analysis into
 * a single comprehensive markdown report.
 */

import { writeFileSync } from "fs";
import { join } from "path";
import * as api from "./api";
import { analyzeSentiment, computeStats, enrichTweets, type EnrichedTweet } from "./sentiment";
import { analyzeQuery, type GrokOpts } from "./grok";
import { trackCost } from "./costs";
import { formatTweetMarkdown } from "./format";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface ReportOpts {
  accounts?: string[];     // additional accounts to include
  sentiment?: boolean;     // run sentiment analysis
  model?: string;          // Grok model for analysis
  pages?: number;          // search pages
  save?: boolean;          // save to file
}

// ---------------------------------------------------------------------------
// Report generation
// ---------------------------------------------------------------------------

export async function generateReport(query: string, opts: ReportOpts = {}): Promise<string> {
  const date = new Date().toISOString().split("T")[0];
  const time = new Date().toISOString().replace("T", " ").slice(0, 19);
  const sections: string[] = [];

  console.error(`Generating report: "${query}"...`);

  // ---- 1. Main topic search ----
  console.error(`  Searching "${query}"...`);
  const tweets = await api.search(query, {
    pages: opts.pages || 2,
    sortOrder: "relevancy",
    since: "1d",
  });
  trackCost("search", "/2/tweets/search/recent", tweets.length);

  const topTweets = api.sortBy(tweets, "likes").slice(0, 20);

  // ---- 2. Account-specific searches ----
  const accountTweets: Record<string, api.Tweet[]> = {};
  if (opts.accounts && opts.accounts.length > 0) {
    for (const acct of opts.accounts) {
      const username = acct.replace(/^@/, "");
      console.error(`  Checking @${username}...`);
      try {
        const { tweets: userTweets } = await api.profile(username, { count: 10 });
        trackCost("profile", `/2/users/by/username/${username}`, userTweets.length + 1);
        accountTweets[username] = userTweets;
      } catch (e: any) {
        console.error(`  Warning: @${username}: ${e.message}`);
      }
    }
  }

  // ---- 3. Sentiment analysis ----
  let sentimentSection = "";
  if (opts.sentiment && topTweets.length > 0) {
    console.error(`  Running sentiment analysis...`);
    try {
      const sentiments = await analyzeSentiment(topTweets, { model: opts.model });
      const stats = computeStats(sentiments);
      const enriched = enrichTweets(topTweets, sentiments);

      sentimentSection += `## Sentiment Analysis\n\n`;
      sentimentSection += `**Average score:** ${stats.average_score.toFixed(2)}/1.0\n`;
      sentimentSection += `| Sentiment | Count | Pct |\n|-----------|-------|-----|\n`;
      const total = topTweets.length;
      sentimentSection += `| Positive | ${stats.positive} | ${Math.round((stats.positive / total) * 100)}% |\n`;
      sentimentSection += `| Negative | ${stats.negative} | ${Math.round((stats.negative / total) * 100)}% |\n`;
      sentimentSection += `| Neutral | ${stats.neutral} | ${Math.round((stats.neutral / total) * 100)}% |\n`;
      if (stats.mixed > 0) {
        sentimentSection += `| Mixed | ${stats.mixed} | ${Math.round((stats.mixed / total) * 100)}% |\n`;
      }
      sentimentSection += "\n";

      // Show tweets with most extreme sentiment
      const positive = enriched.filter(t => t.sentiment?.sentiment === "positive")
        .sort((a, b) => (b.sentiment?.score || 0) - (a.sentiment?.score || 0));
      const negative = enriched.filter(t => t.sentiment?.sentiment === "negative")
        .sort((a, b) => (a.sentiment?.score || 0) - (b.sentiment?.score || 0));

      if (positive.length > 0) {
        sentimentSection += `### Most Positive\n\n`;
        sentimentSection += positive.slice(0, 3).map(formatTweetMarkdown).join("\n\n");
        sentimentSection += "\n\n";
      }
      if (negative.length > 0) {
        sentimentSection += `### Most Negative\n\n`;
        sentimentSection += negative.slice(0, 3).map(formatTweetMarkdown).join("\n\n");
        sentimentSection += "\n\n";
      }
    } catch (e: any) {
      sentimentSection += `## Sentiment Analysis\n\n*Analysis failed: ${e.message}*\n\n`;
    }
  }

  // ---- 4. AI Summary ----
  let aiSummary = "";
  console.error(`  Generating AI summary...`);
  try {
    const tweetContext = topTweets.slice(0, 15)
      .map((t, i) => `[${i + 1}] @${t.username} (${t.metrics.likes}L): ${t.text.slice(0, 280)}`)
      .join("\n");

    const prompt = `Based on these ${topTweets.length} tweets about "${query}", provide:
1. A 2-3 sentence executive summary of the current conversation
2. Key themes (3-5 bullet points)
3. Notable signals or trends
4. Any emerging narratives or sentiment shifts

Be concise and actionable.`;

    const response = await analyzeQuery(prompt, tweetContext, { model: opts.model || "grok-4-1-fast" });
    aiSummary = response.content;
  } catch (e: any) {
    aiSummary = `*AI summary unavailable: ${e.message}*`;
  }

  // ---- Build report ----
  let report = `# Intelligence Report: ${query}\n\n`;
  report += `**Generated:** ${time}\n`;
  report += `**Tweets analyzed:** ${tweets.length}\n`;
  if (opts.accounts && opts.accounts.length > 0) {
    report += `**Tracked accounts:** ${opts.accounts.join(", ")}\n`;
  }
  report += "\n---\n\n";

  // AI Summary
  report += `## Executive Summary\n\n${aiSummary}\n\n`;

  // Sentiment
  if (sentimentSection) {
    report += sentimentSection;
  }

  // Top tweets
  report += `## Top Tweets (by engagement)\n\n`;
  report += topTweets.slice(0, 10).map(formatTweetMarkdown).join("\n\n");
  report += "\n\n";

  // Account sections
  for (const [username, userTweets] of Object.entries(accountTweets)) {
    if (userTweets.length > 0) {
      report += `## @${username} — Recent Activity\n\n`;
      report += userTweets.slice(0, 5).map(formatTweetMarkdown).join("\n\n");
      report += "\n\n";
    }
  }

  // Metadata
  report += `---\n\n## Report Metadata\n\n`;
  report += `- **Query:** ${query}\n`;
  report += `- **Date:** ${date}\n`;
  report += `- **Tweets scanned:** ${tweets.length}\n`;
  report += `- **Est. cost:** ~$${(tweets.length * 0.005).toFixed(2)} (search) + Grok API\n`;
  report += `- **Generated by:** [xint](https://github.com/0xNyk/xint)\n`;

  return report;
}

// ---------------------------------------------------------------------------
// CLI handler
// ---------------------------------------------------------------------------

export async function cmdReport(args: string[]): Promise<void> {
  const queryParts: string[] = [];
  let accounts: string[] = [];
  let sentiment = false;
  let model: string | undefined;
  let pages: number | undefined;
  let save = false;

  let i = 0;
  while (i < args.length) {
    const arg = args[i];
    switch (arg) {
      case "--accounts":
      case "-a": {
        const val = args[++i];
        if (val) accounts = val.split(",").map(a => a.trim());
        break;
      }
      case "--sentiment":
      case "-s":
        sentiment = true;
        break;
      case "--model":
        model = args[++i];
        break;
      case "--pages":
        pages = parseInt(args[++i] || "2");
        break;
      case "--save":
        save = true;
        break;
      case "--help":
      case "-h":
        printReportHelp();
        return;
      default:
        queryParts.push(arg);
    }
    i++;
  }

  const query = queryParts.join(" ");
  if (!query) {
    printReportHelp();
    return;
  }

  const report = await generateReport(query, {
    accounts: accounts.length > 0 ? accounts : undefined,
    sentiment,
    model,
    pages,
    save,
  });

  console.log(report);

  if (save) {
    const slug = query
      .replace(/[^a-zA-Z0-9]+/g, "-")
      .replace(/^-|-$/g, "")
      .slice(0, 40)
      .toLowerCase();
    const date = new Date().toISOString().split("T")[0];
    const SKILL_DIR = import.meta.dir;
    const path = join(SKILL_DIR, "..", "data", "exports", `report-${slug}-${date}.md`);
    writeFileSync(path, report);
    console.error(`\nSaved to ${path}`);
  }
}

function printReportHelp(): void {
  console.log(`
Usage: xint report <topic> [options]

Generate an automated intelligence report combining search results,
sentiment analysis, and AI-powered insights.

Options:
  --accounts, -a <list>  Comma-separated accounts to track (e.g., @user1,@user2)
  --sentiment, -s        Include AI sentiment analysis
  --model <name>         Grok model: grok-4, grok-4-1-fast (default), grok-3, grok-3-mini
  --pages <N>            Search pages, 1-5 (default: 2)
  --save                 Save report to data/exports/

Examples:
  xint report "AI agents"
  xint report "solana" --sentiment --save
  xint report "react 19" --accounts @reactjs,@dan_abramov --sentiment
  xint report "crypto market" -a @VitalikButerin,@caborek -s --model grok-3
`);
}

File: .skills/openclaw-skills/skills/0xnyk/xint-rs/src/commands/report.rs
use anyhow::Result;
use std::fs;

use crate::api::{grok, twitter};
use crate::cli::ReportArgs;
use crate::client::XClient;
use crate::config::Config;
use crate::costs;
use crate::format;
use crate::models::*;
use crate::sentiment;

pub async fn run(args: &ReportArgs, config: &Config, client: &XClient) -> Result<()> {
    let token = config.require_bearer_token()?;
    let query = args.topic.join(" ");

    if query.is_empty() {
        println!("Usage: xint report <topic> [options]");
        return Ok(());
    }

    let time = chrono::Utc::now().format("%Y-%m-%d %H:%M:%S").to_string();
    let date = &time[..10];

    eprintln!("Generating report: \"{query}\"...");

    // 1. Main topic search
    eprintln!("  Searching \"{query}\"...");
    let tweets = twitter::search(
        client,
        token,
        &query,
        args.pages,
        "relevancy",
        Some("1d"),
        None,
        false,
    )
    .await?;

    costs::track_cost(
        &config.costs_path(),
        "search",
        "/2/tweets/search/recent",
        tweets.len() as u64,
    );

    let mut top_tweets = tweets.clone();
    twitter::sort_by(&mut top_tweets, "likes");
    top_tweets.truncate(20);

    // 2. Account-specific searches
    let accounts: Vec<String> = args
        .accounts
        .as_ref()
        .map(|a| {
            a.split(',')
                .map(|s| s.trim().trim_start_matches('@').to_string())
                .collect()
        })
        .unwrap_or_default();

    let mut account_sections = Vec::new();
    for acct in &accounts {
        eprintln!("  Checking @{acct}...");
        match twitter::get_profile(client, token, acct, 10, false).await {
            Ok((_user, user_tweets)) => {
                costs::track_cost(
                    &config.costs_path(),
                    "profile",
                    &format!("/2/users/by/username/{acct}"),
                    user_tweets.len() as u64 + 1,
                );
                account_sections.push((acct.clone(), user_tweets));
            }
            Err(e) => {
                eprintln!("  Warning: @{acct}: {e}");
            }
        }
    }

    // 3. Sentiment analysis
    let mut sentiment_section = String::new();
    if args.sentiment && !top_tweets.is_empty() {
        if let Ok(api_key) = config.require_xai_key() {
            eprintln!("  Running sentiment analysis...");
            let http = reqwest::Client::new();
            match sentiment::analyze_sentiment(
                &http,
                api_key,
                &top_tweets,
                Some(&args.model),
                Some(&config.costs_path()),
            )
            .await
            {
                Ok(sentiments) => {
                    let stats = sentiment::compute_stats(&sentiments);
                    let total = top_tweets.len();
                    let pct = |n: u32| -> u32 {
                        if total > 0 {
                            (n as f64 / total as f64 * 100.0).round() as u32
                        } else {
                            0
                        }
                    };

                    sentiment_section.push_str("## Sentiment Analysis\n\n");
                    sentiment_section.push_str(&format!(
                        "**Average score:** {:.2}/1.0\n",
                        stats.average_score
                    ));
                    sentiment_section
                        .push_str("| Sentiment | Count | Pct |\n|-----------|-------|-----|\n");
                    sentiment_section.push_str(&format!(
                        "| Positive | {} | {}% |\n",
                        stats.positive,
                        pct(stats.positive)
                    ));
                    sentiment_section.push_str(&format!(
                        "| Negative | {} | {}% |\n",
                        stats.negative,
                        pct(stats.negative)
                    ));
                    sentiment_section.push_str(&format!(
                        "| Neutral | {} | {}% |\n",
                        stats.neutral,
                        pct(stats.neutral)
                    ));
                    if stats.mixed > 0 {
                        sentiment_section.push_str(&format!(
                            "| Mixed | {} | {}% |\n",
                            stats.mixed,
                            pct(stats.mixed)
                        ));
                    }
                    sentiment_section.push('\n');
                }
                Err(e) => {
                    sentiment_section.push_str(&format!(
                        "## Sentiment Analysis\n\n*Analysis failed: {e}*\n\n"
                    ));
                }
            }
        }
    }

    // 4. AI Summary
    eprintln!("  Generating AI summary...");
    let ai_summary;
    if let Ok(api_key) = config.require_xai_key() {
        let http = reqwest::Client::new();
        let tweet_context =
            grok::format_tweets_for_context(&top_tweets[..top_tweets.len().min(15)]);

        let prompt = format!(
            "Based on these {} tweets about \"{}\", provide:\n\
             1. A 2-3 sentence executive summary of the current conversation\n\
             2. Key themes (3-5 bullet points)\n\
             3. Notable signals or trends\n\
             4. Any emerging narratives or sentiment shifts\n\n\
             Be concise and actionable.",
            top_tweets.len(),
            query
        );

        let opts = GrokOpts {
            model: args.model.clone(),
            ..Default::default()
        };

        match grok::analyze_query_tracked(
            &http,
            api_key,
            &prompt,
            Some(&tweet_context),
            &opts,
            Some(&config.costs_path()),
        )
        .await
        {
            Ok(response) => ai_summary = response.content,
            Err(e) => ai_summary = format!("*AI summary unavailable: {e}*"),
        }
    } else {
        ai_summary = "*AI summary unavailable: XAI_API_KEY not set*".to_string();
    }

    // Build report
    let mut report = format!("# Intelligence Report: {query}\n\n");
    report.push_str(&format!("**Generated:** {time}\n"));
    report.push_str(&format!("**Tweets analyzed:** {}\n", tweets.len()));
    if !accounts.is_empty() {
        report.push_str(&format!(
            "**Tracked accounts:** {}\n",
            accounts
                .iter()
                .map(|a| format!("@{a}"))
                .collect::<Vec<_>>()
                .join(", ")
        ));
    }
    report.push_str("\n---\n\n");

    report.push_str(&format!("## Executive Summary\n\n{ai_summary}\n\n"));

    if !sentiment_section.is_empty() {
        report.push_str(&sentiment_section);
    }

    report.push_str("## Top Tweets (by engagement)\n\n");
    for t in top_tweets.iter().take(10) {
        report.push_str(&format::format_tweet_markdown(t));
        report.push_str("\n\n");
    }

    for (username, user_tweets) in &account_sections {
        if !user_tweets.is_empty() {
            report.push_str(&format!("## @{username} — Recent Activity\n\n"));
            for t in user_tweets.iter().take(5) {
                report.push_str(&format::format_tweet_markdown(t));
                report.push_str("\n\n");
            }
        }
    }

    report.push_str("---\n\n## Report Metadata\n\n");
    report.push_str(&format!("- **Query:** {query}\n"));
    report.push_str(&format!("- **Date:** {date}\n"));
    report.push_str(&format!("- **Tweets scanned:** {}\n", tweets.len()));
    report.push_str(&format!(
        "- **Est. cost:** ~${:.2} (search) + Grok API\n",
        tweets.len() as f64 * 0.005
    ));
    report.push_str("- **Generated by:** [xint](https://github.com/0xNyk/xint)\n");

    println!("{report}");

    if args.save {
        let exports_dir = config.exports_dir();
        fs::create_dir_all(&exports_dir)?;
        let slug = query
            .chars()
            .filter(|c| c.is_alphanumeric() || *c == ' ')
            .collect::<String>()
            .replace(' ', "-")
            .to_lowercase();
        let slug = &slug[..slug.len().min(40)];
        let path = exports_dir.join(format!("report-{slug}-{date}.md"));
        fs::write(&path, &report)?;
        eprintln!("\nSaved to {}", path.display());
    }

    Ok(())
}
